In [32]:
import numpy as np 
from sklearn.tree import DecisionTreeClassifier, export_graphviz
import graphviz

In [33]:
f = open('test_add.dat')
d = []
for l in f:
  d.append([float(x) for x in l.split()])
d = np.array(d)

x = d[:, :-1]
y = d[:, -1]
x, y

(array([[105., 452.,  26.,  57.],
        [748., 386., 415., 836.],
        [938.,   4., 668., 221.],
        ...,
        [372., 930., 852., 107.],
        [957., 643., 307., 350.],
        [232., 226., 437., 611.]], shape=(1000000, 4)),
 array([0., 1., 1., ..., 0., 1., 0.], shape=(1000000,)))

In [34]:
depth = np.arange(1, 30010, 1000)
min_leaf = [0.0] #Этот параметр всегда будет уменьшать score, поэтому оставляю 0.0
max_node = np.arange(2, 30010, 1000)


best_score = 0.0
best_param = None
best_clf = None

high_scores = []
high_scores_params = []

for i in depth:
    for j in min_leaf:
        for k in max_node:
            clf = DecisionTreeClassifier(max_depth=i, min_weight_fraction_leaf=j, max_leaf_nodes=k)
            clf.fit(x, y)
            score = clf.score(x, y)
            if score > 0.998:
              high_scores.append(score)
              high_scores_params.append((i, j, k))
            
            print(f"Пробую параметры: depth={i}, leaf={j}, nodes={k}. Результат: {score}")

            if best_score < score:
                best_score = score
                best_param = (i, j, k)
                best_clf = clf
                print(f"score изменился - {best_score}")
                
print("Оценки > 0.998:", high_scores)
print("Параметры для этих оценок:", high_scores_params)

Пробую параметры: depth=1, leaf=0.0, nodes=2. Результат: 0.926269
score изменился - 0.926269
Пробую параметры: depth=1, leaf=0.0, nodes=1002. Результат: 0.926269
Пробую параметры: depth=1, leaf=0.0, nodes=2002. Результат: 0.926269
Пробую параметры: depth=1, leaf=0.0, nodes=3002. Результат: 0.926269
Пробую параметры: depth=1, leaf=0.0, nodes=4002. Результат: 0.926269
Пробую параметры: depth=1, leaf=0.0, nodes=5002. Результат: 0.926269
Пробую параметры: depth=1, leaf=0.0, nodes=6002. Результат: 0.926269
Пробую параметры: depth=1, leaf=0.0, nodes=7002. Результат: 0.926269
Пробую параметры: depth=1, leaf=0.0, nodes=8002. Результат: 0.926269
Пробую параметры: depth=1, leaf=0.0, nodes=9002. Результат: 0.926269
Пробую параметры: depth=1, leaf=0.0, nodes=10002. Результат: 0.926269
Пробую параметры: depth=1, leaf=0.0, nodes=11002. Результат: 0.926269
Пробую параметры: depth=1, leaf=0.0, nodes=12002. Результат: 0.926269
Пробую параметры: depth=1, leaf=0.0, nodes=13002. Результат: 0.926269
Пробую

In [35]:
print(high_scores)
print(high_scores_params)
#По выводу данных мы можем сделать вывод 
#Что максимально score мы досгиаем благодаря nodes >= 30000 при глубине >=1000
#Поэтому лучшую модель мы будем строить с учетом этого

[0.998503, 0.999036, 0.998525, 0.999064, 0.998519, 0.999056, 0.998515, 0.999053, 0.998514, 0.999043, 0.998516, 0.999053, 0.99853, 0.999055, 0.998527, 0.999038, 0.998512, 0.999055, 0.998502, 0.999059, 0.99852, 0.999042, 0.998534, 0.999052, 0.998513, 0.999056, 0.99853, 0.999067, 0.998523, 0.999054, 0.998517, 0.999037, 0.998536, 0.999049, 0.998518, 0.999038, 0.998518, 0.999056, 0.998513, 0.999041, 0.998536, 0.999051, 0.998531, 0.999051, 0.998497, 0.999041, 0.998507, 0.99905, 0.998517, 0.999035, 0.998521, 0.999054, 0.998531, 0.999062, 0.998524, 0.999064, 0.998519, 0.999053, 0.998517, 0.999047]
[(np.int64(1001), 0.0, np.int64(29002)), (np.int64(1001), 0.0, np.int64(30002)), (np.int64(2001), 0.0, np.int64(29002)), (np.int64(2001), 0.0, np.int64(30002)), (np.int64(3001), 0.0, np.int64(29002)), (np.int64(3001), 0.0, np.int64(30002)), (np.int64(4001), 0.0, np.int64(29002)), (np.int64(4001), 0.0, np.int64(30002)), (np.int64(5001), 0.0, np.int64(29002)), (np.int64(5001), 0.0, np.int64(30002)), (n

In [36]:
# Доказательство того что leaf уменьшает общий score
leaf = np.arange(0.1, 0.51, 0.1)
for i in leaf:
  clf = DecisionTreeClassifier(max_depth = 1000, max_leaf_nodes = 30000, min_weight_fraction_leaf = i)
  clf.fit(x, y)
  score = clf.score(x, y)
  print(f"Пробую параметры: depth={1000}, leaf={i}, nodes={30000}. Результат: {score}")
  


Пробую параметры: depth=1000, leaf=0.1, nodes=30000. Результат: 0.926269
Пробую параметры: depth=1000, leaf=0.2, nodes=30000. Результат: 0.926269
Пробую параметры: depth=1000, leaf=0.30000000000000004, nodes=30000. Результат: 0.926269
Пробую параметры: depth=1000, leaf=0.4, nodes=30000. Результат: 0.926269
Пробую параметры: depth=1000, leaf=0.5, nodes=30000. Результат: 0.575573


In [37]:

clf = DecisionTreeClassifier(max_depth=1000, max_leaf_nodes=30000)
clf.fit(x, y)
clf.score(x, y)
#Подобранны наилучшие параметры

0.999059

### С учетом, того что лучший скор достигается путем изменение только одного параметра(Такой вывод был сделан на основе анализа вывода данных), можно предположить что можно достичь скора при изменении максимальной глубины

In [38]:
depth = np.arange(1, 1000, 10)
best_depth = 0
best_score = 0
for i in depth:
  clf = DecisionTreeClassifier(max_depth=i)
  clf.fit(x, y)
  score = clf.score(x, y)
  if score == 1:
    best_score = score
    best_depth = i
    break
print(f"Score равный единице найден - {best_score}")
print(f"Score равен одну при max_depth = {best_depth}")

Score равный единице найден - 1.0
Score равен одну при max_depth = 71


In [39]:
max_leaf = np.arange(2, 30100, 1000)
best_score = 0
best_leaf = 0
for i in max_leaf:
  clf = DecisionTreeClassifier(max_leaf_nodes=i)
  clf.fit(x, y)
  score = clf.score(x, y)
  if score >= 0.999:
    best_score = score
    best_leaf = i
    print(f"Найден score >= 0.999 - {best_score}")
    print(f"Score достигает значения {best_score} при max_leaf_nodes = {best_leaf}")


Найден score >= 0.999 - 0.999044
Score достигает значения 0.999044 при max_leaf_nodes = 30002


### На основе прошлой гипотезы предпологаю так же что только при изменении min_weight_fraction_leaf score тоже будет близок к единице

In [40]:
leaf = np.arange(0.1, 0.5, 0.1)
best_score = 0
best_weight = 0
for i in leaf:
  clf = DecisionTreeClassifier(min_weight_fraction_leaf=i)
  clf.fit(x, y)
  score = clf.score(x, y)
  if score == 1:
    best_score = score
    best_weight = i
if best_score <= 0.998:
  print("Результат не найден")
else:
    print(f"Найден score >= 0.999 - {best_score}")
    print(f"Score достигает значения {best_score} при min_weight_fraction_leaf = {best_weight}")
# С изменением минимального единицы никогда не достигнет 1
# Моя гипотеза неверна
    

Результат не найден


# Вывод

Задача по подбору параметров для достижения точности > 0.999 успешно выполнена. В ходе экспериментов были найдены следующие варианты ограничений дерева:

1.  **Ограничение только по максимальной глубине (`max_depth`):**
    - При `max_depth=71` достигается идеальная точность **1.0**.

2.  **Ограничение только по максимальному числу листовых узлов (`max_leaf_nodes`):**
    - При `max_leaf_nodes=30002` достигается точность **0.999081**.

3.  **Ограничение по `min_weight_fraction_leaf`:**
    - Экспериментально доказано, что для максимальной точности параметр `min_weight_fraction_leaf` должен быть равен **0.0**. Любое его увеличение снижает точность на данном наборе данных.
4. **Ограничение по `max_depth` и `max_lef nodes`**
    - При max_depth >=1000 и max_leaf_nodes >=30000 достигается точность **0.999032**
